### PR TITLE
[Cpp]use global state accessor to get session dir, if necessary

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
## Why are these changes needed?
Follow Java and Python we can reuse `GetInternalKV` in global state accessor to get `session_dir` so that we can remove `hiredis` dependency.
This PR also optimized the getting logic, let it do only if necessary.

## Related issue number

First PR of https://github.com/ray-project/ray/issues/15357.
Let's see if there's more to do.
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: cpp/src/ray/util/process_helper.cc [hunk 1]
```diff
@@ -100,13 +83,14 @@ void ProcessHelper::RayStart(CoreWorkerOptions::TaskExecutionCallback callback)
     }
   }
 
+  std::unique_ptr<ray::gcs::GlobalStateAccessor> global_state_accessor = nullptr;
   if (ConfigInternal::Instance().worker_type == ray::WorkerType::DRIVER) {
-    ray::gcs::GlobalStateAccessor global_state_accessor(
-        redis_address, ConfigInternal::Instance().redis_password);
-    RAY_CHECK(global_state_accessor.Connect()) << "Failed to connect to GCS.";
+    global_state_accessor.reset(new ray::gcs::GlobalStateAccessor(
```
Comment (qicosmos) [inline cpp/src/ray/util/process_helper.cc:88]: Use an object is better than a pointer here.

File: cpp/src/ray/util/process_helper.cc [hunk 2]
```diff
@@ -119,14 +103,18 @@ void ProcessHelper::RayStart(CoreWorkerOptions::TaskExecutionCallback callback)
...
 
-  auto session_dir = ConfigInternal::Instance().session_dir.empty()
-                         ? GetSessionDir(redis_ip, ConfigInternal::Instance().redis_port,
-                                         ConfigInternal::Instance().redis_password)
-                         : ConfigInternal::Instance().session_dir;
-
-  auto log_dir = ConfigInternal::Instance().logs_dir.empty()
-                     ? session_dir + "/logs"
-                     : ConfigInternal::Instance().logs_dir;
+  std::string log_dir = ConfigInternal::Instance().logs_dir;
```
Comment (qicosmos) [inline cpp/src/ray/util/process_helper.cc:106]: the log_dir is a temporary, it can be moved later.

File: cpp/src/ray/util/process_helper.cc [hunk 3]
```diff
@@ -100,13 +83,14 @@ void ProcessHelper::RayStart(CoreWorkerOptions::TaskExecutionCallback callback)
     }
   }
 
+  std::unique_ptr<ray::gcs::GlobalStateAccessor> global_state_accessor = nullptr;
```
Comment (jovany-wang) [inline cpp/src/ray/util/process_helper.cc:86]: Why we don't initial GlobalStateAccessor here?

**Review by jovany-wang — COMMENTED** [review]
Looks good to me overall. Let us wait for cpp team guys to take a review.

File: cpp/src/ray/util/process_helper.cc (hunk 3 above)
Comment (WangTaoTheTonic) [reply cpp/src/ray/util/process_helper.cc:86]: We don't always need this accessor. For example if session_dir is passed in command line in worker, we don't need it.

File: cpp/src/ray/util/process_helper.cc (hunk 1 above)
Comment (WangTaoTheTonic) [reply cpp/src/ray/util/process_helper.cc:88]: We don't always init this object so a pointer is convenient for check if it's initialized.

File: cpp/src/ray/util/process_helper.cc (hunk 2 above)
Comment (WangTaoTheTonic) [reply cpp/src/ray/util/process_helper.cc:106]: ok

**Review by qicosmos — APPROVED** [review]

**Comment by jovany-wang** [comment]
I believe these failures are unrelated because this PR only changes cpp worker. merging.


```bash
python/ray/util/dask:dask_ray_shuffle_optimization_client_mode

(03:28:02) [41 / 45] 6 / 30 tests, 2 failed; Testing //rllib:test_impala_pong_deterministic_v4_40k_ts_1G_obj_store; 41s local, remote-cache ... (5 actions running)

  | (03:28:57) [41 / 45] 6 / 30 tests, 2 failed; Testing //rllib:test_impala_pong_deterministic_v4_40k_ts_1G_obj_store; 96s local, remote-cache ... (5 actions running)
  | (03:30:00) [49 / 53] 8 / 30 tests, 2 failed; Testing //rllib:test_impala_pong_deterministic_v4_40k_ts_1G_obj_store; 160s local, remote-cache ... (5 actions running)
  | (03:31:13) [57 / 61] 10 / 30 tests, 2 failed; Testing //rllib:test_impala_pong_deterministic_v4_40k_ts_1G_obj_store; 233s local, remote-cache ... (6 actions running)
  | (03:32:37) [61 / 65] 11 / 30 tests, 2 failed; Testing //rllib:test_dqn_pong_deterministic_v4; 90s local ... (6 actions running)
```

**Pull request merged**